pkg/credentialprovider/plugin/config.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/criauth"
+)
+
+// readCRIAuthPluginConfigFile receives a path to a config file and decodes it
+// into the internal CRIAuthPluginConfig type.
+func readCRIAuthPluginConfigFile(configPath string) (*criauth.CRIAuthPluginConfig, error) {
+	if configPath == "" {
+		return nil, fmt.Errorf("CRI auth plugin config path is empty")
+	}
+
+	data, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read external registry credential provider configuration from %q: %v", configPath, err)
+	}
+
+	config, err := decode(data)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding config %s: %v", configPath, err)
+	}
+
+	return config, nil
+}
+
+// decode decodes data into the internal CRIAuthPluginConfig type.
+func decode(data []byte) (*criauth.CRIAuthPluginConfig, error) {
+	obj, gvk, err := codecs.UniversalDecoder().Decode(data, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if gvk.Kind != "CRIAuthPluginConfig" {
+		return nil, fmt.Errorf("failed to decode %q (wrong Kind)", gvk.Kind)
+	}
+
+	if internalConfig, ok := obj.(*criauth.CRIAuthPluginConfig); ok {
+		return internalConfig, nil
+	}
+
+	return nil, fmt.Errorf("unable to convert %T to *CRIAuthPluginConfig", obj)
+}
+
+func validateCRIAuthPluginConfig(config *criauth.CRIAuthPluginConfig) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(config.Plugins) == 0 {
+		allErrs = append(allErrs, field.Required(field.NewPath("plugins"), "at least 1 item in plugins is required"))
+	}
+
+	fieldPath := field.NewPath("plugins")
+	for _, plugin := range config.Plugins {
+		if strings.Contains(plugin.Name, ".") || strings.Contains(plugin.Name, "/") || strings.Contains(plugin.Name, " ") {
+			allErrs = append(allErrs, field.Invalid(fieldPath.Child("name"), plugin.Name, "plugin name cannot contain '.' or '/' or spaces"))
+		}
+
+		if len(plugin.MatchImages) == 0 {
+			allErrs = append(allErrs, field.Required(fieldPath.Child("matchImages"), "at least 1 item in matchUrls is required"))
```

Comment:
Should the error message be `at least one item in matchImages is required`?

Reply:
Good catch, will update